br/pkg/utils/retry.go

```diff
@@ -84,3 +87,76 @@ func FallBack2CreateTable(err error) bool {
 	}
 	return false
 }
+
+// RetryWithBackoffer is a simple context for a "mixed" retry.
+// Some of TiDB APIs, say, `ResolveLock` requires a `tikv.Backoffer` as argument.
+// But the `tikv.Backoffer` isn't pretty customizable, it has some sorts of predefined configuration but
+// we cannot create new one. So we are going to mix up the flavour of `tikv.Backoffer` and our homemade
+// back off strategy. That is what the `RetryWithBackoffer` did.
+type RetryWithBackoffer struct {
+	bo *tikv.Backoffer
+
+	totalBackoff int
+	maxBackoff   int
+	baseErr      error
+
+	mu          sync.Mutex
+	nextBackoff int
+}
+
+// AdaptTiKVBackoffer creates an "ad-hoc" backoffer, which wraps a backoffer and provides some new functions:
+// When backing off, we can manually provide it a specified sleep duration instead of directly provide a retry.Config
+// Which is sealed in the "client-go/internal".
+func AdaptTiKVBackoffer(ctx context.Context, maxSleepMs int, baseErr error) *RetryWithBackoffer {
+	return &RetryWithBackoffer{
+		bo:         tikv.NewBackoffer(ctx, maxSleepMs),
+		maxBackoff: maxSleepMs,
+		baseErr:    baseErr,
+	}
+}
+
+// NextSleepInMS returns the time `BackOff` will sleep in ms of the state.
+func (r *RetryWithBackoffer) NextSleepInMS() int {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.nextBackoff
+}
+
+// TotalSleepInMS returns the total sleeped time in ms.
+func (r *RetryWithBackoffer) TotalSleepInMS() int {
+	return r.totalBackoff + r.bo.GetTotalSleep()
```

Comment:
It seems that the `r.bo` is always forked, and only the forked `bo`s are updated something. So is `r.bo.GetTotalSleep` always 0?

Reply:
Anyway unlocking locks has a bounded backoff time, which is the TTL of the lock. I think omit the retry time it cost won't be really harmful, hopefully...

Reply:
The sole difference between the original logic is that resolve lock may fail due to the accumulated sleep time reaches the sleep time limit. I think that would be fine to eliminate that fail case.